Refactor the loading state of the sign on forms

tgolen opened the pull request:
### Details
This refactors the forms to use an Onyx state property rather than their individual internal state

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/149339
Fixes https://github.com/Expensify/Expensify.cash/issues/1133
Fixes https://github.com/Expensify/Expensify.cash/issues/1131

### Tests
1. Make sure you're signed out of the app
2. Sign in with an existing account
3. Verify the form buttons in the sign-in flow have a proper loading state
3. Verify that each step except for the login step has a link for "Change Expensify login" and that clicking the link takes you back to the first step
3. Sign out
4. Sign in with a brand new account and verify the form buttons in the sign-in flow have a proper loading state
5. Sign out
6. Enter an email with an existing account and click continue
7. Navigate to `/#/setpassword/test`
8. Enter the password `123` and submit the form
9. Verify the button shows a brief loading state and there is an error that the password is invalid
10. Enter the password `asdfASDF00` and submit the form
11. Verify there is an error message saying the validate code is likely expired (that's OK, because we used a fake validate code in the URL)

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
<Add any necessary screenshots for all platforms if your change added or updated UI>

#### Web
no changes

#### Mobile Web
no changes

#### Desktop
no changes

#### iOS
no changes

#### Android
no changes

tgolen committed:
Refactor the login form to use session.isLoading

tgolen committed:
Refactor the password form to use session.isLoading

tgolen committed:
Refactor the githubUsername form to use session.isLoading

tgolen committed:
Refactor the validation link form to use session.isLoading

tgolen committed:
Add a default setting for the isLoading prop

tgolen committed:
Move default setting for isLoading onto the button component

tgolen committed:
Remove unused imports

marcaaron commented on ios/Podfile.lock:
```diff
@@ -672,4 +672,4 @@ SPEC CHECKSUMS:
 
 PODFILE CHECKSUM: 96882cac17249b4e6e01afb5d10b1762a5191341
 
-COCOAPODS: 1.10.0
```
Any reason this is changing? Seems like maybe people are using different versions of cocoapods which one is right?

sketchydroide replied:
Agree, also seem to be downgrading, which I think there would be no reason to.

tgolen replied:
Ah, I keep forgetting to stop committing this. Thanks for catching it

marcaaron commented on src/pages/signin/SubmitButton.js:
```diff
@@ -3,80 +3,63 @@ import PropTypes from 'prop-types';
-    const isLoading = props.isLoading && !props.session.error;
-    return (
-        <>
+const SubmitButton = props => (
```
Suggestion to rename this to something like `SignInControls` or `LoginButtonGroup` a "submit button" sounds more general purpose and the presence of "Change Expensify login" tells me this is more specific than the name would suggest.

tgolen replied:
Ah, that's a great point. I had a similar thought when this component was tied to the `session` property in Onyx. I would like to move the "change expensify login" to be outside the button control, and that will allow this to be more generic and maybe move it to the components folder.

marcaaron commented on src/libs/actions/Session.js:
```diff
@@ -96,6 +96,9 @@ function fetchAccountDetails(login) {
             }
+        })
+        .finally(() => {
+            Onyx.merge(ONYXKEYS.SESSION, {isLoading: false});
```
This works, but I just want to make sure my original suggestion was known here. i.e. that we should consider linking the loading states with the data keys we are waiting to populate.

e.g. if we are calling `GetAccountStatus` then it would make sense to tie the loading state into `ONYXKEYS.ACCOUNT`. Just an idea. Not sure if it's super valuable to do now. But e.g. then we could skip this call to `finally()` and instead do:

```js
Onyx.merge(ONYXKEYS.ACCOUNT, { 
    loading: false,
    ...params,
});
```

Yeah this does mean we'd have to update all the places where this is used to be more specific instead of reusing `session` as a general loading state tracker, but I think it's a good pattern that if a component is calling an API it will be subscribing to the data it modifies as well.

tgolen replied:
OK, I think that makes sense. Are you proposing that there would be both `account.isLoading` and `session.isLoading`? Is that all or would there also be `credentials.isLoading`?

marcaaron replied:
Yep all of them. In other words, we are waiting for some _data_ to load so the key where we will eventually set the data will hold the `loading` state.

tgolen replied:
OK, so it would work like this:

1. Login form - `account.loading`
2. GitHub username form - this needs both `account` and `credentials`
3. Password form - also needs both `account` and `credentials`
4. Resend validation link form - also needs both `account` and `credentials`

Since three of them need access to two pieces of Onyx data, and it's the same data needed for every step, I'm not sure what Onyx key would be used for those cases.

marcaaron replied:
Good question and conversation. I guess I'm not sure how to answer. I want to say whatever data model you are updating with data returned from the API response would be what you toggle the `loading` state on. But I guess there could be multiple APIs being called and different data being set so maybe this isn't the best solution!

We are sort of replacing the traditional "make API request, handle response, do something" flow with something where the "done" state of the action is hooked into some random-ish piece of Onyx state. 

Going high level with this for a second, but what if every action had it's own meta data key in Onyx and we set these loading states _there_ instead of tying them to components (which seems bad for reusability) or sticking them on pieces of data (which maybe don't need loading states at all)

tgolen replied:
> We are sort of replacing the traditional "make API request, handle response, do something" flow with something where the "done" state of the action is hooked into some random-ish piece of Onyx state.

I think this is a good thing to get clarity on. The way I see it, and also is the vision behind what @cead22 is attempting in https://github.com/Expensify/Expensify.cash/pull/915 is that basically the entire app is a reflection of what's in Onyx. There is a spectrum between nothing in Onyx (you are signed out and not looking at anything) to Onyx being fully loaded (you are looking at reports and adding actions). The UI just follows along on this journey that essentially is all about getting the right data into Onyx. 

So yes, it's definitely different than how we would traditionally approach this. I think that's why just using a single `session.loading` is fine because the entire sign on process is all about populating `session.authToken`.

I'm not really a big fan of letting actions store meta-data in their own Onyx keys because I think that can be easily abused. What happens when enough of that meta data uses up the storage limits and reports can't be stored anymore? I know that's extreme, but it could eventually happen.

I think that if we don't use `session.isLoading`, then I'd prefer to go back to using state in the individual forms (that code was more simple and used less lines to do the same thing). One thing that I've seen with testing is that the `isLoading` can persist between page refreshes if you refresh at the right time (or if you refresh when an API request is hanging) which actually makes having it be in Onyx a little bit worse in my opinion.

marcaaron replied:
>which actually makes having it be in Onyx a little bit worse in my opinion

Interesting take, but that does sound like a possible side effect of the data journey the UI is on. 

Is the persistence of the loading spinner because we are redirecting before the loading state has updated? If we redirect once our action resolves (rather than determining which view to display based on the state of our various props) then maybe we could avoid this? e.g.

- Make request A
- Handle response A
- Set loading state to false
- Set the correct view after the action is done

I'll admit, that doesn't fit the approach you are describing of a UI that is going on a data journey. But this is what the journey looks like now...
https://github.com/Expensify/Expensify.cash/blob/bf4926b93d884aa41d942a3e351485931cc5ebbc/src/pages/signin/SignInPage.js#L70-L106

Others might disagree, but this logic is somewhat difficult for me to parse.

Anyways, I won't block on these changes if you are OK with leaving the question of how to handle a loading state unresolved for now. I do think these types of questions will pop up in the future. I don't think `Onyx` provides clear guidance on how to tackle this yet. But it should eventually!

Maybe we can consider investigating / taking some inspiration from redux thunks in this case.
https://redux.js.org/tutorials/essentials/part-5-async-logic#loading-state-for-requests

tgolen replied:
> Is the persistence of the loading spinner because we are redirecting before the loading state has updated? If we redirect once our action resolves

It doesn't have anything to do with redirects, no. There aren't any redirects in this flow at all. The situation is like this:

1. Submit the LoginForm
2. `isLoading` is now `true` and spinner on the button is shown
3. The API takes a long time to respond, maybe a couple of seconds
4. User gets impatient and refreshes the page
5. `isLoading` is still `true` so there is a spinner on the button even though nothing is loading

I think this kind of goes back to a convo I remember having with Rory about what is persisted to Onyx and what is not. I think one of the outcomes of that is that we should only put things on Onyx that we want to still be there after the page is reloaded. The open/closed state of the hamburger is a good example of something we want to be remembered after a refresh. I don't think loading spinners (especially ones associated with API requests that are canceled when the page is refreshed) are something that should be persisted.

This has me more convinced that this should be returned to the original implementation of using the state of the forms.

tgolen replied:
After making that comment, I had another thought... 

Maybe we should have loading states automatically reset when the app loads. That could solve the problem I posted above.

As I looked into that, I came across the code in `Expensify.js` for `Onyx.init()` which actually does exactly that. Only it's clearing out `session.loading` and my code here is using `session.isLoading`.

Now there is a bigger issue with my code in this PR as those two properties are very confusing and not consistent that I need to think through.

marcaaron reviewed (commented):
This looks good! I'm OK with merging this as is - but I'm am not 100% on using the `session` key in a generic way to store the loading state. More in comments.

sketchydroide commented:
Appart from Marcs comments everything looks good to me, I'll be happy to approve once they are resolved :)

tgolen committed:
Remove pod lockfile

tgolen committed:
Remove the change expensify login link to it's own component

tgolen committed:
Rename the button component to be generic and move it to the component folder

tgolen committed:
Import react

tgolen commented:
I've pushed an update here to make the button a generic component. I didn't change anything with the loading keys yet since that's still an open conversation.

tgolen committed:
Merge branch 'master' into tgolen-refactor-loading

tgolen committed:
Move the loading state to the account key

tgolen committed:
Move error and loading to the account key and clean up set password

tgolen committed:
Remove unused imports

tgolen committed:
Fix the display name

tgolen committed:
Make this logic more consistent with other forms

tgolen commented:
OK, I've made another pretty big update to this PR. I've moved the `error` and `loading` off of `SESSION` and put it onto `ACCOUNT` so that it doesn't interfere with the other places in the code using `SESSION.loading` (eg. the router).

I've also updated the `SetPassword` page to use the new code changes so that it's consistent with the rest.

marcaaron commented on src/libs/actions/Session.js:
```diff
@@ -96,6 +96,10 @@ function fetchAccountDetails(login) {
+            Onyx.merge(ONYXKEYS.ACCOUNT, {error: response.message});
+        })
+        .finally(() => {
+            Onyx.merge(ONYXKEYS.ACCOUNT, {loading: false});
```
NAB, might be able to combine these into a single merge in the `finally()` by declaring a 

```js
const account = {};
```
before the API call and then modified in `.then()`

Not sure if that's much of an improvement but I thought it could be an interesting convention to follow

marcaaron commented on src/libs/actions/Session.js:
```diff
@@ -218,7 +234,10 @@ function setPassword(password, validateCode) {
+            Onyx.merge(ONYXKEYS.ACCOUNT, {error: response.message});
+        })
+        .finally(() => {
+            Onyx.merge(ONYXKEYS.ACCOUNT, {loading: false});
```
Just thinking out loud and not requesting a change. But seeing all of this logic makes me wonder if it might be valuable to create a higher order function or system that automatically applies this loading state logic to any API command by passing a related Onyx key. But not sure if we need that yet.

marcaaron commented on src/pages/signin/ResendValidationForm.js:
```diff
@@ -12,10 +31,15 @@ class ResendValidationForm extends React.Component {
 
+    componentWillUnmount() {
+        if (this.successMessageTimer) {
+            clearInterval(this.successMessageTimer);
```
This should be a `clearTimeout()`

marcaaron commented on src/pages/signin/ResendValidationForm.js:
```diff
@@ -40,14 +64,15 @@ class ResendValidationForm extends React.Component {
                 </View>
 
-                {this.state.formSuccess && (
+                {!_.isEmpty(this.state.formSuccess) && (
```
Just curious, but why did this need to be changed from the string check? 

I _thought_ I saw something about how `empty string && render something` will try to render the empty string on native platforms which is bad.

marcaaron replied:
Yes that must be it. I wonder if there is a way to get the linter to enforce this... 🤔

marcaaron replied:
Maybe just switch to ternaries for everything and ban short circuit 🤷

tgolen replied:
Yeah, this is specific to using string values and `&&` which breaks mobile. It would be nice to enforce this with a linter somehow, but it's not possible today. I'd hate to lose the ability to use the `&&` shortcut because ternarys make the code much less clean IMO (eg. seeing `: null` everywhere is tiresome and noisy).

marcaaron reviewed (changes requested):
Looks great to me just one tiny comment and a few random thoughts / questions (not change requests).

tgolen committed:
Use cleartimeout instead

tgolen commented:
Updated to fix clearTimeout. Thanks for spotting that. I think we should go ahead and get this bad boy merged. Thanks for the other thoughts!
